syncer/data_fetch.go

```diff
@@ -0,0 +1,342 @@
+			logger.With().Warning("failed fetching new blocks",
+				log.Array("block_ids", log.ArrayMarshalerFunc(func(encoder log.ArrayEncoder) error {
+					for _, bid := range blocksToFetch {
+						encoder.AppendString(bid.String())
+					}
+					return nil
+				})),
+				log.Err(err))
+			// syntactically invalid blocks are expected from malicious peers
+		}
+	}
+}
+
+// PollLayerOpinions polls all peers for opinions in the specified layer.
+func (d *DataFetch) PollLayerOpinions(ctx context.Context, lid types.LayerID) ([]*fetch.LayerOpinion, error) {
+	peers := d.fetcher.GetPeers()
+	if len(peers) == 0 {
+		return nil, errNoPeers
+	}
+	req := &opinionRequest{
+		lid:   lid,
+		peers: peers,
+		ch:    make(chan peerResult[fetch.LayerOpinion], len(peers)),
+	}
+	okFunc := func(data []byte, peer p2p.Peer) {
+		d.receiveOpinions(ctx, req, peer, data, nil)
+	}
+	errFunc := func(err error, peer p2p.Peer) {
+		d.receiveOpinions(ctx, req, peer, nil, err)
+	}
+	if err := d.fetcher.GetLayerOpinions(ctx, peers, lid, okFunc, errFunc); err != nil {
+		return nil, err
+	}
+	req.peerResults = map[p2p.Peer]peerResult[fetch.LayerOpinion]{}
+	var (
+		success      bool
+		candidateErr error
+	)
+	for {
+		select {
+		case res := <-req.ch:
+			req.peerResults[res.peer] = res
+			if res.err == nil {
+				success = true
+				req.response.opinions = append(req.response.opinions, res.data)
+			} else if candidateErr == nil {
+				candidateErr = res.err
+			}
+			if len(req.peerResults) < len(req.peers) {
+				break
+			}
+			// all peer responded
+			if !success {
+				req.response.err = candidateErr
+			}
+			return req.response.opinions, req.response.err
+		case <-ctx.Done():
+			return nil, errTimeout
+		}
+	}
+}
+
+func (d *DataFetch) receiveOpinions(ctx context.Context, req *opinionRequest, peer p2p.Peer, data []byte, peerErr error) {
+	logger := d.logger.WithContext(ctx).WithFields(req.lid, log.Stringer("peer", peer))
+	logger.Debug("received layer opinions from peer")
+
+	var (
+		result = peerResult[fetch.LayerOpinion]{peer: peer, err: peerErr}
+		lo     fetch.LayerOpinion
+	)
+	if peerErr != nil {
+		logger.With().Debug("received peer error for layer opinions", log.Err(peerErr))
+	} else if result.err = codec.Decode(data, &lo); result.err != nil {
+		logger.With().Debug("error converting bytes to LayerOpinion", log.Err(result.err))
+	} else {
+		lo.SetPeer(peer)
+		result.data = &lo
+	}
+	select {
+	case req.ch <- result:
+	case <-ctx.Done():
+	}
+}
+
+type epochAtxRes struct {
+	err    error
+	atxIDs []types.ATXID
+}
+
+func randomPeer(peers []p2p.Peer) p2p.Peer {
+	idx := rand.Intn(len(peers))
+	return peers[idx]
+}
+
+// GetEpochATXs fetches all ATXs in the specified epoch from a peer.
+func (d *DataFetch) GetEpochATXs(ctx context.Context, epoch types.EpochID) error {
+	resCh := make(chan epochAtxRes, 1)
+	peers := d.fetcher.GetPeers()
+	if len(peers) == 0 {
+		return errNoPeers
+	}
+	peer := randomPeer(peers)
```

Comment:
Imo there is nothing wrong with writing

```suggestion
	peer := peers[rand.Intn(len(peers))]
```

instead of having `randomPeer` function that is only used in one place.